htdocs/web_portal/components/Get_User_Principle.php

```diff
@@ -158,11 +158,23 @@ function Get_User_Principle(){
         // If cannot find user, search certificate DNs instead
         if ($user === null) {
             $user = $serv->getUserByCertificateDn($principleString);
+            $authExists = False;
+        } else {
+            $authExists = True;
+        }
 
         // Is user registered/known in the DB? if true, update their last login time
         // once for the current request.
         if ($user !== null) {
             $serv->updateLastLoginTime($user);
+
+            // If property for current auth does not exist, add to user
+            if (!$authExists) {
+                // Get type of auth logged in with e.g. IGTF)
+                $authType = $auth->getDetails()['AuthenticationRealm'][0];
+                $propArr = array($authType, $principleString);
+                $serv->addUserProperty($user, $propArr, $user);
```

Comment:
After setting up the User Property, shouldn't we also set the certificate DN to some unique placeholder value?

Reply:
There aren't immediate problems with having the ID string stored in both places, which is why at this point I hadn't yet overwritten the certificate DN. When it is overwritten, it's done within `addUserProperty`, so this section of the code wouldn't actually change.

I implemented the overwrite from the [registration](https://github.com/GOCDB/gocdb/pull/294/commits/6d03e14a073bc439b1a671789427aa235f3cfc68) onwards, as that was the point when setting the certificate DN was actually necessary, although the final form requires the [schema change](https://github.com/GOCDB/gocdb/pull/294/commits/3c75b41c5ef9a41adb80aa321adbc26bb4a48f92) (which is a separate commit, as you requested).

Reply:
It would be clearer if the overwriting of the old certificateDN field was factored out into it's own method and called here.

Reply:
Thinking about this more, calling an overwrite method here would come with downsides to consider.

The first is hopefully fairly minor: the same function would have to be called in other parts of the code (e.g. during linking/recovery with an old account), which increases the risk that it isn't called when needed. This isn't really a reason against doing it, more something to be aware/careful of.

More significantly, calling it here would mean that overwriting certificateDn and adding the user property would not be wrapped in a single transaction, unless we added that here too. Overwriting could be done after the user property has been successfully added, which would be less risky, but then validation that the ID string is unique would be made more complicated, as it checks against certificateDns as well as other user properties.

Reply:
So we definitely want overwriting `certificateDn` and adding the User Identity wrapped in a single transaction, as we don't want a User to get into a state where they have both a valid `certificateDn` entry and a User Identity (or worse they have neither of those things)

maybe we need a `migrateUserCredentials` method (for example) for this case to wrap the logic of `addUserPropertyLogic` and "overwrite certificateDn" into a single transaction.

Reply:
This was done.